hamt/hamt.go

```diff
@@ -400,6 +435,40 @@ func (ds *Shard) ForEachLink(ctx context.Context, f func(*ipld.Link) error) erro
 	})
 }
 
+// makeAsyncTrieGetLinks builds a getLinks function that can be used with EnumerateChildrenAsync
+// to iterate a HAMT shard. It takes an IPLD Dag Service to fetch nodes, and a call back that will get called
+// on all links to leaf nodes in a HAMT tree, so they can be collected for an EnumLinks operation
+func makeAsyncTrieGetLinks(dagService ipld.DAGService, onShardValue func(*shardValue) error) dag.GetLinks {
+
+	return func(ctx context.Context, currentCid cid.Cid) ([]*ipld.Link, error) {
+		node, err := dagService.Get(ctx, currentCid)
+		if err != nil {
+			return nil, err
+		}
+		directoryShard, err := NewHamtFromDag(dagService, node)
+		if err != nil {
+			return nil, err
+		}
+
+		childShards := make([]*ipld.Link, 0, len(directoryShard.children))
+		for idx := range directoryShard.children {
+			lnk := directoryShard.nd.Links()[idx]
```

Comment:
Just so it doesn't get lost in the noise, we need to figure how to walk the trie, relying on the links at the DAG layer isn't enough since they don't always represent the current state of the shard. An approach similar to `Node()` should be used.

Reply:
Sorry, you're right, I got confused with the `range` over the uninitialized `children` slice, but it is allocated with a space equal to the node's links, so the `range` over the indexes should work.